compiler/vm/cgen2.nim

```diff
@@ -0,0 +1,2450 @@
+  else:
+    let name = getTypeName(gen.cache, env, id, Declaration())
+    var decl = genCTypeDecl(gen, id)
+
+    # set the identifier field for struct and union types:
+    if decl[0].kind in {cdnkStruct, cdnkUnion}:
+      decl[1] = (cdnkIdent, name.uint32, 0'u32)
+
+    result = CTypeInfo(decl: decl, name: name)
+
+
+func useFunction(c: var ModuleCtx, s: ProcId) =
+  c.funcs.incl s
+
+func useType(c: var ModuleCtx, t: TypeId) =
+  assert t != NoneType
+  c.types.incl t
+
+type GenCtx = object
+  sym: ProcId
+
+  names: seq[CAst] # IRIndex -> expr
+  exprs: seq[ExprInfo]
+  types: seq[TypeId]
+  config: ConfigRef
+
+  scope: NameScope
+  localNames: seq[CIdent]
+
+  env: #[lent]# ptr IrEnv
+
+  gl: GlobalGenCtx # XXX: temporary
+  m: ModuleCtx # XXX: temporary
+
+func getUniqueName(scope: NameScope, cache: var IdentCache, name: string): CIdent =
+  ## Returns a name that is not already in use in the given `scope`. If the
+  ## given `name` is not in use, the corresponding identifier ID is returned.
+  ## Otherwise, `name` appended to with a number is returned.
+  ## Does **not** register the returned identifier with `scope`.
+
+  result = cache.getOrIncl(name)
+
+  if scope.idents.contains(result):
+    # an identifier with the requested name already exists
+    var buf = name
+    buf.add "_"
+    let origLen = buf.len
+
+    var next = 10
+    var i = 0
+    while true:
+      while i < next:
+        buf.setLen(origLen)
+        buf.addInt(i)
+        result = cache.getOrIncl(buf)
+        if not scope.idents.contains(result):
+          return
+
+        inc i
+
+      # no free name was found in the previous range. Expand the search range.
+      # Not resetting `i` back to 0 means that we're wasting character space,
+      # but it also prevents `a_1` and `a_01` from both existing in the same
+      # scope
+
+      next *= 10
+
+func gen(c: GenCtx, irs: IrStore3, n: IRIndex): CAst =
+  c.names[n]
+
+func mapTypeV3(t: TypeId): CTypeId
+
+func mapTypeV2(c: var GenCtx, t: TypeId): CTypeId =
+  # TODO: unfinished
+  c.m.useType(t) # mark the type as used
+  mapTypeV3(t)
+
+func mapTypeV3(t: TypeId): CTypeId =
+  if t != NoneType:
+    # TODO: no ``NoneType`` should reach here
+    # XXX: maybe just have a ``NoneType`` -> ``VoidCType`` mapping in the table instead?
+    CTypeId(t)
+  else:
+    VoidCType
+
+func genCProcHeader(idents: var IdentCache, env: ProcedureEnv, s: ProcId): CProcHeader =
+  result.ident = idents.getOrIncl(mangledName(env, s))
+  result.returnType = mapTypeV3(env.getReturnType(s))
+
+  result.args.newSeq(env.numParams(s))
+  var i = 0
+  for p in env.params(s):
+    let ident =
+      if p.name != nil: idents.getOrIncl(mangledName(p.name))
+      else:             idents.getOrIncl("_param" & $i)
+
+    result.args[i] = (mapTypeV3(p.typ),
+                      ident)
+    inc i
+
+
+template start(): CAstBuilder =
+  var b: CAstBuilder
+  b
+
+template buildAst(code: untyped): CAst =
```

Comment:
I think
`start().....fin()` or `fin: start()....` is simpler than `buildAst: builder....`.